File: algorithms4th/2.3/quickSort.py
def partition(data, lo, hi):
    v = data[lo]
    i = lo
    j = hi + 1

    while i < j:
        while True:
            i += 1
            if data[i] > v or i >= hi:
                break
        while True:
            j -= 1
            if data[j] < v or j <= lo:
                break
        if i >= j:
            break

        data[i], data[j] = data[j], data[i]

    data[j], data[lo] = data[lo], data[j]
    return j

def quickSort(data, lo, hi):
    if lo >= hi:
        return
    mid = partition(data, lo, hi)
    quickSort(data, lo, mid - 1)
    quickSort(data, mid + 1, hi)

data = [0, 0, 0, 9, 9, 9, 8, 4, 1, 3, 5, 2, 6, 7]
#data = [1, 1, 1, 1]

quickSort(data, 0, len(data) - 1)
print(data)

File: ahaAlgorithms/quickSort.cpp
//
// Created by liuyang on 2019-01-07.
//

#include<vector>
#include<cstdio>

using std::vector;

int partation(vector<int> &nums, int left, int right);

void quickSort(vector<int> &nums, int left, int right) {
    if (left > right) {
        return;
    }

    int mid = partation(nums, left, right);
    quickSort(nums, left, mid - 1);
    quickSort(nums, mid + 1, right);
}

int partation(vector<int> &nums, int left, int right) {

    int target = nums[left];

    int i = left;
    int j = right;

    while(i < j) {
        for (j = right; i < j && nums[j] >= target; --j) {}
        for (i = left; i < j && nums[i] <= target; ++i) {}
        if (i < j) {
            std::swap(nums[i], nums[j]);
        }
    }
    std::swap(nums[left], nums[i]);
    return i;
}

int main() {
    vector<int> data = {6, 1, 2, 7, 9, 3, 4, 5, 10, 8};
    quickSort(data, 0, int(data.size() - 1));
    for(auto i : data) {
        printf("%d ", i);
    }
}
